src/python/pants/backend/go/target_type_rules.py

```diff
@@ -24,8 +46,103 @@ async def inject_go_module_dependency(request: InjectGoModuleDependency) -> Inje
+class InferGoPackageDependenciesRequest(InferDependenciesRequest):
+    infer_from = GoPackageSources
+
+
+@rule
+async def infer_go_dependencies(
+    request: InferGoPackageDependenciesRequest,
+    goroot_imports: ResolvedImportPathsForGoLangDistribution,
+) -> InferredDependencies:
+    this_go_package = await Get(
+        ResolvedGoPackage, ResolveGoPackageRequest(request.sources_field.address)
+    )
+
+    # Obtain all go_package targets under this package's go_module.
+    assert this_go_package.module_address is not None
+    spec_path = this_go_package.module_address.spec_path
+    address_specs = [
+        MaybeEmptySiblingAddresses(spec_path),
+        MaybeEmptyDescendantAddresses(spec_path),
+    ]
+    candidate_targets = await Get(UnexpandedTargets, AddressSpecs(address_specs))
+    go_package_targets = [
+        tgt
+        for tgt in candidate_targets
+        if tgt.has_field(GoPackageSources)
+        and not tgt.address.is_file_target
+        and tgt.address != this_go_package.address
+    ]
+
+    # Find all go_external_modules in the repo and map their import paths to their address.
+    candidate_go_external_module_targets = await Get(
+        UnexpandedTargets, AddressSpecs([MaybeEmptyDescendantAddresses("")])
+    )
+    go_external_module_targets = [
+        tgt for tgt in candidate_go_external_module_targets if isinstance(tgt, GoExternalModule)
+    ]
+    third_party_import_path_to_address = {
+        tgt[GoImportPath].value: tgt.address for tgt in go_external_module_targets
+    }
+
+    # Resolve all of the packages found.
+    first_party_import_path_to_address = {}
+    first_party_go_packages = await MultiGet(
+        Get(ResolvedGoPackage, ResolveGoPackageRequest(tgt.address)) for tgt in go_package_targets
+    )
+    for first_party_go_package in first_party_go_packages:
+        # Skip packages that are not part of this package's module.
+        # TODO: This requires that all first-party code in the monorepo be part of the same go_module. Will need
+        # figure out how multiple modules in a monorepo can interact.
+        if first_party_go_package.module_address != this_go_package.module_address:
+            continue
+
+        first_party_import_path_to_address[
+            first_party_go_package.import_path
+        ] = first_party_go_package.address
+
+    # Loop through all of the imports of this package and add dependencies on other packages and
+    # external modules.
+    inferred_dependencies = []
+    for import_path in this_go_package.imports + this_go_package.test_imports:
+        # Check whether the import path comes from the standard library.
+        if import_path in goroot_imports.import_path_mapping:
+            continue
+
+        # Infer first-party dependencies to other packages in same go_module.
+        if import_path in first_party_import_path_to_address:
+            inferred_dependencies.append(first_party_import_path_to_address[import_path])
+            continue
```

Comment:
Is it possible to have ambiguity here, where something is both first and third party, and needs disambiguiation? If so, this should probably not `continue`, and instead collect all matches.

Reply:
Yes if the external dependencies are vendored (which won't be supported for a while in any case). I'll add a check if the import path is present in both places.

Reply:
Would be good to wait for @Eric-Arellano to take a look as well: it looks like the other inference rules request `ExplicitlyProvidedDependencies` in order to execute their disambiguation, and he might have an idea of how we can avoid each set of inference rules needing to do that.

Reply:
Yeah, as explained above, handling ambiguity will be really important (fine in followup). I suspect there are 3 sources of ambiguity:

- \>1 first party target w/ same import path
- \>1 third party target w/ same import path (is this possible in Go?)
- first party and third party refer to same import path

The Python code in `dependency_inference/` is a good example of how to handle this. Basically, the import path mappings will keep track of which import paths are ambiguous. They don't eagerly warn, only keep track. Then, your rule to map a particular Go import path to its target will keep track of ambiguous vs. unambiguous owners:

https://github.com/pantsbuild/pants/blob/650a9b3cd26bc48751e2c9e570699b6779dbfc0b/src/python/pants/backend/python/dependency_inference/module_mapper.py#L318-L384

Finally, you'd use this code to try to disambiguate and then warn:

https://github.com/pantsbuild/pants/blob/650a9b3cd26bc48751e2c9e570699b6779dbfc0b/src/python/pants/backend/python/dependency_inference/rules.py#L149-L165 

There's some added complexity in that we allow for Python to sometimes have 2 owners of the same module when using type stubs. You don't need that complexity, sounds like there should either be 1 owner or it's ambiguous.